File: pascal/concepts/arrays.pas
unit Arrays;

interface

procedure DemonstrateArrays;

implementation

procedure DemonstrateArrays;
const
  MaxSize = 4;
type
  NumberArray = array[1..MaxSize] of Integer;
  FruitArray = array[1..MaxSize] of string;
var
  Numbers: NumberArray;
  Fruits: FruitArray;
  I: Integer;
begin
  WriteLn;
  WriteLn('=== Arrays ===');
  
  { Initialize arrays }
  for I := 1 to MaxSize do
    Numbers[I] := I * 10;
    
  Fruits[1] := 'Apple';
  Fruits[2] := 'Banana';
  Fruits[3] := 'Orange';
  Fruits[4] := 'Grape';
  
  { Display number array }
  WriteLn('Number array:');
  for I := 1 to MaxSize do
    Write(Numbers[I], ' ');
  WriteLn;
  
  { Display fruit array }
  WriteLn;
  WriteLn('Fruit array:');
  WriteLn('First fruit: ', Fruits[1]);
  WriteLn('Last fruit: ', Fruits[MaxSize]);
  WriteLn('All fruits:');
  for I := 1 to MaxSize do
    WriteLn('  ', I, ': ', Fruits[I]);
end;

end.

File: rust/concepts/arrays.rs
pub fn demonstrate_arrays() {
    println!("\n=== Arrays & Vectors ===");

    // An array has a fixed size, defined at compile time.
    let fixed_size_array: [&str; 3] = ["apple", "banana", "orange"];
    println!("Fixed-size array: {:?}", fixed_size_array);

    // A Vector is a growable, heap-allocated list, similar to a JS array or Go slice.
    let fruits_vec = vec!["apple", "banana", "orange", "grape"];
    println!("Vector (growable): {:?}", fruits_vec);

    // Access the first element.
    println!("First fruit of vector: {}", fruits_vec[0]);

    // Access the last element using the `.len()` method.
    // We subtract 1 because indices are 0-based.
    println!("Last fruit of vector: {}", fruits_vec[fruits_vec.len() - 1]);
}
